File: api/internal/model/vip.go
package model

// VipLevel VIP等级配置（完全可配置）
type VipLevel struct {
	Base
	Slug        string `json:"slug" gorm:"size:32;uniqueIndex"`                           // 等级标识（可配置，如 vip1/vip2/svip1）
	Title       string `json:"title" gorm:"size:64;not null"`                             // 等级名称（显示用）
	ExpRequired int64  `json:"expRequired" gorm:"column:exp_required;not null;default:0"` // 升级所需累计消费/经验（分）

	// 永久折扣
	OrderDiscount float64 `json:"orderDiscount" gorm:"column:order_discount;default:1.0"` // 下单永久折扣 (0.98 = 98折, 1.0 = 无折扣)

	// 月度券配置（绑定优惠券模板）
	MonthlyCouponTemplateID *uint64 `json:"monthlyCouponTemplateId" gorm:"column:monthly_coupon_template_id;index"` // 月度券模板ID
	MonthlyCouponCount      int     `json:"monthlyCouponCount" gorm:"column:monthly_coupon_count;default:0"`        // 每月发放数量

	// 展示配置
	IconURL   string `json:"iconUrl" gorm:"column:icon_url;size:255"`      // 等级图标
	Color     string `json:"color" gorm:"size:32"`                         // 等级颜色（前端展示）
	Benefits  string `json:"benefits" gorm:"type:json;default:'{}'"`       // 其他权益描述(JSON)
	SortOrder int    `json:"sortOrder" gorm:"column:sort_order;default:0"` // 排序（越小越靠前）

	// 状态
	IsDefault bool `json:"isDefault" gorm:"column:is_default;default:false"` // 是否默认等级（新用户解锁后）
	IsActive  bool `json:"isActive" gorm:"column:is_active;default:true"`    // 是否启用
}

// TableName 指定表名
func (VipLevel) TableName() string {
	return "vip_levels"
}

// HasOrderDiscount 是否有永久折扣
func (v *VipLevel) HasOrderDiscount() bool {
	return v.OrderDiscount > 0 && v.OrderDiscount < 1.0
}

// HasMonthlyCoupon 是否有月度券
func (v *VipLevel) HasMonthlyCoupon() bool {
	return v.MonthlyCouponTemplateID != nil && v.MonthlyCouponCount > 0
}

// VipConfig VIP系统配置（全局配置）
type VipConfig struct {
	Base
	ConfigKey   string `json:"configKey" gorm:"column:config_key;size:64;uniqueIndex"` // 配置键
	ConfigValue string `json:"configValue" gorm:"column:config_value;type:text"`       // 配置值
	Description string `json:"description" gorm:"size:255"`                            // 描述
}

// TableName 指定表名
func (VipConfig) TableName() string {
	return "vip_configs"
}

// VIP配置键常量
const (
	VipConfigUnlockByConsume  = "unlock_by_consume"  // 累计消费解锁门槛（分）
	VipConfigUnlockByRecharge = "unlock_by_recharge" // 累计充值解锁门槛（分）
	VipConfigExpireDays       = "expire_days"        // VIP过期天数（0=永久）
)

File: admin/src/api/vip.ts
import apiClient from './client';
import type { ApiResponse } from './admin';

// ============================================================================
// VIP Level Types
// ============================================================================

interface VIPLevel {
    id: number;
    slug: string;
    title: string;
    expRequired: number;
    orderDiscount: number;
    monthlyCouponTemplateId?: number;
    monthlyCouponCount: number;
    iconUrl: string;
    color: string;
    benefits: string;
    sortOrder: number;
    isDefault: boolean;
    isActive: boolean;
    createdAt: string;
    updatedAt: string;
}

interface CreateVIPLevelDto {
    slug: string;
    title: string;
    expRequired?: number;
    orderDiscount?: number;
    monthlyCouponTemplateId?: number;
    monthlyCouponCount?: number;
    iconUrl?: string;
    color?: string;
    benefits?: string;
    sortOrder?: number;
    isDefault?: boolean;
    isActive?: boolean;
}

interface UpdateVIPLevelDto {
    slug: string;
    title: string;
    expRequired?: number;
    orderDiscount?: number;
    monthlyCouponTemplateId?: number;
    monthlyCouponCount?: number;
    iconUrl?: string;
    color?: string;
    benefits?: string;
    sortOrder?: number;
    isDefault?: boolean;
    isActive?: boolean;
}

interface VIPLevelQueryParams {
    page?: number;
    page_size?: number;
    keyword?: string;
    isActive?: boolean;
}

// ============================================================================
// VIP Config Types
// ============================================================================

interface VIPConfig {
    id: number;
    configKey: string;
    configValue: string;
    description: string;
    createdAt: string;
    updatedAt: string;
}

interface CreateVIPConfigDto {
    configKey: string;
    configValue: string;
    description?: string;
}

interface UpdateVIPConfigDto {
    configValue: string;
    description?: string;
}

// ============================================================================
// VIP Config Key Constants
// ============================================================================

export const VIP_CONFIG_KEYS = {
    UNLOCK_BY_CONSUME: 'unlock_by_consume',  // 累计消费解锁门槛（分）
    UNLOCK_BY_RECHARGE: 'unlock_by_recharge', // 累计充值解锁门槛（分）
    EXPIRE_DAYS: 'expire_days',              // VIP过期天数（0=永久）
} as const;

export type VIPConfigKey = typeof VIP_CONFIG_KEYS[keyof typeof VIP_CONFIG_KEYS];

// ============================================================================
// Batch Operation Types
// ============================================================================

interface VIPBatchUpdateStatusDto {
    ids: number[];
    isActive: boolean;
}

interface VIPBatchDeleteDto {
    ids: number[];
}

// ============================================================================
// VIP API
// ============================================================================

export const vipApi = {
    // ------------------------------------------------------------------------
    // VIP Level Management
    // ------------------------------------------------------------------------

    /**
     * Get VIP levels list with pagination
     * GET /api/v1/admin/vip/levels
     */
    getVIPLevels: (params?: VIPLevelQueryParams) =>
        apiClient.get<ApiResponse<VIPLevel[]>>('/admin/vip/levels', { params }),

    /**
     * Get VIP level detail by ID
     * GET /api/v1/admin/vip/levels/:id
     */
    getVIPLevelDetail: (id: number) =>
        apiClient.get<ApiResponse<VIPLevel>>(`/admin/vip/levels/${id}`),

    /**
     * Create a new VIP level
     * POST /api/v1/admin/vip/levels
     */
    createVIPLevel: (data: CreateVIPLevelDto) =>
        apiClient.post<ApiResponse<VIPLevel>>('/admin/vip/levels', data),

    /**
     * Update VIP level by ID
     * PUT /api/v1/admin/vip/levels/:id
     */
    updateVIPLevel: (id: number, data: UpdateVIPLevelDto) =>
        apiClient.put<ApiResponse<VIPLevel>>(`/admin/vip/levels/${id}`, data),

    /**
     * Delete VIP level by ID
     * DELETE /api/v1/admin/vip/levels/:id
     */
    deleteVIPLevel: (id: number) =>
        apiClient.delete<ApiResponse<void>>(`/admin/vip/levels/${id}`),

    /**
     * Set default VIP level
     * POST /api/v1/admin/vip/levels/:id/default
     */
    setDefaultVIPLevel: (id: number) =>
        apiClient.post<ApiResponse<void>>(`/admin/vip/levels/${id}/default`),

    /**
     * Batch update VIP level status
     * POST /api/v1/admin/vip/levels/batch-status
     */
    batchUpdateVIPLevelStatus: (data: VIPBatchUpdateStatusDto) =>
        apiClient.post<ApiResponse<{ affected: number }>>('/admin/vip/levels/batch-status', data),

    /**
     * Batch delete VIP levels
     * POST /api/v1/admin/vip/levels/batch-delete
     */
    batchDeleteVIPLevels: (data: VIPBatchDeleteDto) =>
        apiClient.post<ApiResponse<{ affected: number }>>('/admin/vip/levels/batch-delete', data),

    // ------------------------------------------------------------------------
    // VIP Config Management
    // ------------------------------------------------------------------------

    /**
     * Get all VIP configs
     * GET /api/v1/admin/vip/configs
     */
    getVIPConfigs: () =>
        apiClient.get<ApiResponse<VIPConfig[]>>('/admin/vip/configs'),

    /**
     * Get VIP config by key
     * GET /api/v1/admin/vip/configs/:key
     */
    getVIPConfig: (key: string) =>
        apiClient.get<ApiResponse<VIPConfig>>(`/admin/vip/configs/${key}`),

    /**
     * Create or update VIP config
     * POST /api/v1/admin/vip/configs
     */
    saveVIPConfig: (data: CreateVIPConfigDto) =>
        apiClient.post<ApiResponse<VIPConfig>>('/admin/vip/configs', data),

    /**
     * Update VIP config by key
     * PUT /api/v1/admin/vip/configs/:key
     */
    updateVIPConfig: (key: string, data: UpdateVIPConfigDto) =>
        apiClient.put<ApiResponse<VIPConfig>>(`/admin/vip/configs/${key}`, data),

    /**
     * Delete VIP config by key
     * DELETE /api/v1/admin/vip/configs/:key
     */
    deleteVIPConfig: (key: string) =>
        apiClient.delete<ApiResponse<void>>(`/admin/vip/configs/${key}`),
};

// ============================================================================
// Re-exports
// ============================================================================

export type {
    VIPLevel,
    CreateVIPLevelDto,
    UpdateVIPLevelDto,
    VIPLevelQueryParams,
    VIPConfig,
    CreateVIPConfigDto,
    UpdateVIPConfigDto,
    VIPBatchUpdateStatusDto,
    VIPBatchDeleteDto,
};
